### PR TITLE
Rewriting push event handling; allowing data to be optional.

### LINKED ISSUE #161: Make `PushEvent.data` nullable
Per discussion with @martinthomson, once encryption lands, we'll want a way to distinguish messages without an entity body from encrypted, blank messages. For the former, it seems sensible to set `PushEvent.data` to null.

### PR DESCRIPTION
**Pull request by martinthomson**
This is a follow-on to #164.  The event algorithm needed cleaning up, and while I was there, I had to mark PushEvent.data nullable.

Closes #161.

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -816,36 +816,56 @@
           </dt>
         </dl>
         <p>
-          Upon receiving a <a>push message</a> for a <a>push subscription</a> from the <a>push
-          service</a> the <a>user agent</a> MUST run the following steps:
+          Upon receiving a <a>push message</a> from the <a>push service</a> the <a>user agent</a>
```
Comment (beverloo) [inline index.html]: micro nit: having a comma after `push service` makes the sentence a bit easier to read.

File: index.html [hunk 2]
```diff
@@ -816,36 +816,56 @@
...
+          following steps:
+            <ol>
+              <li>Set <var>global</var> to the global object that was provided as an argument.
+              </li>
+              <li>Create a <a>trusted event</a>, <var>e</var>, that uses the
+              <code><a>PushMessageEvent</a></code> interface, with the event type
+              <code>push</code>, which does not bubble, is not cancelable, and has no default
+              action.
+              </li>
+              <li>If the <a>push message</a> contains valid data, set the <code>data</code>
```
Comment (beverloo) [inline index.html]: What does _valid data_ mean? I don't want to fire `onpush` for messages that have undecryptable payloads.

**Comment by beverloo** [comment]
This looks good to me. Thanks!

File: index.html [hunk 3]
```diff
@@ -816,36 +816,56 @@
...
-          events</a> MUST be modified as follows: instead of setting the <code>data</code>
-          attribute of the event to the value of the <var>eventInitDict</var>'s "<code id=
-          "widl-PushEventInit-data">data</code>" member, set the <code>data</code> attribute to a
-          new <a>PushMessageData</a> with <a>bytes</a> set to the result of <a data-lt=
-          "extract a byte sequence">extracting a byte sequence</a> from that dictionary member, or
-          an empty byte sequence if the "<code id="widl-PushEventInit-data">data</code>" member is
-          not provided.
+          events</a> are modified so that instead of setting the <code>data</code> attribute of the
+          event to the value of the <var>eventInitDict</var>'s "<code>data</code>" member, set the
+          <code>data</code> attribute to a new <a>PushMessageData</a> instances with <a>bytes</a>
```
Comment (beverloo) [inline index.html]: instances -> instance

In general, this sentence is very hard to parse.

**Comment by martinthomson** [comment]
Rebased.

**Comment by mvano** [comment]
Looks good to me, my apologies for the slow response. Did you still want to make this change @martinthomson? Also @beverloo: do we need to make changes for this in Chrome for M49 or can it wait until M50?

**Comment by martinthomson** [comment]
Rebased and ready to merge.

**Comment by beverloo** [comment]
We'll need to update Chrome, but that can be done in the M50 timeframe.
